// carousel-angular/esm2020/lib/carousel.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CarouselComponent } from './carousel.component';
import * as i0 from "@angular/core";
export class IvyCarouselModule {
}
IvyCarouselModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: IvyCarouselModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
IvyCarouselModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.2.12", ngImport: i0, type: IvyCarouselModule, declarations: [CarouselComponent], imports: [CommonModule], exports: [CarouselComponent] });
IvyCarouselModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: IvyCarouselModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: IvyCarouselModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [CarouselComponent],
                    imports: [CommonModule],
                    exports: [CarouselComponent],
                    providers: [],
                    bootstrap: [],
                    entryComponents: [CarouselComponent]
                }]
        }] });
