// carousel-angular/esm2020/lib/carousel.mjs
export class Carousel {
    constructor(properties, utils, cells, container, slide) {
        this.properties = properties;
        this.utils = utils;
        this.cells = cells;
        this.container = container;
        this.slide = slide;
        this.autoplayId = null;
        this.handleTouchstart = (event) => {
            this.container.handleTouchstart();
            this.slide.handleTouchstart(event);
        };
        this.handleHorizontalSwipe = (event) => {
            this.container.handleHorizontalSwipe();
        };
        this.handleTouchend = (event) => {
            if (this.properties.freeScroll) {
                this.container.handleTouchend();
            }
            else {
                this.container.handleTouchend(true);
                this.slide.handleTouchend(event);
            }
        };
        this.isNextArrowDisabled = () => this.slide.isNextArrowDisabled();
        this.isPrevArrowDisabled = () => this.slide.isPrevArrowDisabled();
        this.init();
    }
    get cellLength() {
        return this.cells.cellLength;
    }
    get lastCellIndex() {
        return this.cells.cellLength - 1;
    }
    get overflowCellsLimit() {
        return this.utils.overflowCellsLimit;
    }
    get autoplayIsPossible() {
        return this.properties.autoplayIsPossible;
    }
    get margin() {
        return this.properties.margin;
    }
    get minSwipeDistance() {
        return this.properties.minSwipeDistance;
    }
    get transitionDuration() {
        return this.properties.transitionDuration;
    }
    get transitionTimingFunction() {
        return this.properties.transitionTimingFunction;
    }
    get fullCellWidth() {
        return this.properties.cellWidth + this.margin;
    }
    get numberOfVisibleCells() {
        return this.utils.numberOfVisibleCells;
    }
    get slideCounter() {
        return this.slide.counter;
    }
    updateProperties(properties) {
        this.properties = properties;
    }
    init() {
        this.cellsElement = this.properties.cellsElement;
        this.visibleWidth =
            this.properties.visibleWidth ||
                this.cellsElement.parentElement.clientWidth;
    }
    destroy() {
        clearInterval(this.autoplayId);
    }
    lineUpCells() {
        this.cells.lineUp();
    }
    handleTransitionend() {
        this.slide.handleTransitionend();
    }
    next(length = 1) {
        if (!this.isNextArrowDisabled()) {
            this.slide.next(length);
        }
    }
    prev(length = 1) {
        this.slide.prev(length);
    }
    autoplay() {
        if (this.autoplayId === null) {
            this.autoplayId = setInterval(() => {
                if (this.autoplayIsPossible) {
                    this.next();
                }
            }, this.properties.autoplayInterval);
        }
    }
    stopAutoplay() {
        if (this.autoplayId != null) {
            clearInterval(this.autoplayId);
            this.autoplayId = null;
        }
    }
}
