// carousel-angular/esm2020/lib/cells.mjs
export class Cells {
    constructor(carouselProperties, utils) {
        this.carouselProperties = carouselProperties;
        this.utils = utils;
        this.counter = 0;
        this.init(carouselProperties);
    }
    get cellLength() {
        return this.cells ? this.cells.length : 0;
    }
    get fullCellWidth() {
        return this.carouselProperties.cellWidth + this.carouselProperties.margin;
    }
    get cellLengthInLightDOMMode() {
        return this.cellLength;
    }
    get numberOfVisibleCells() {
        return this.utils.numberOfVisibleCells;
    }
    get overflowCellsLimit() {
        return this.utils.overflowCellsLimit;
    }
    updateProperties(carouselProperties) {
        this.carouselProperties = carouselProperties;
    }
    lineUp() {
        const cells = this.element ? this.element.children : [];
        for (let i = 0; i < cells.length; i++) {
            const cell = cells[i];
            const positionX = this.getCellPositionInContainer(i);
            cell.style.transform = `translateX(${positionX}px)`;
            cell.style.width = `${this.carouselProperties.cellWidth}px`;
        }
    }
    ifSequenceOfCellsIsChanged() {
        const cells = this.element.children;
        return cells[0].style.transform !== 'translateX(0px)';
    }
    getCellPositionInContainer(cellIndexInDOMTree) {
        return cellIndexInDOMTree * this.fullCellWidth;
    }
    setCounter(value) {
        this.counter = value;
    }
    init(carouselProperties) {
        this.element = this.carouselProperties.cellsElement;
        this.cells = this.element.children;
        this.visibleWidth =
            this.carouselProperties.visibleWidth ||
                this.element.parentElement.clientWidth;
    }
}
