// carousel-angular/esm2020/lib/utils.mjs
export class Utils {
    constructor(carouselProperties) {
        this.carouselProperties = carouselProperties;
    }
    get margin() {
        return this.carouselProperties.margin;
    }
    get overflowCellsLimit() {
        return this.carouselProperties.overflowCellsLimit;
    }
    get numberOfVisibleCells() {
        return Math.ceil(this.visibleWidth / this.fullCellWidth);
    }
    get visibleCellsOverflowContainer() {
        return (this.numberOfVisibleCells * this.fullCellWidth - this.margin >
            this.visibleWidth);
    }
    get fullCellWidth() {
        return this.carouselProperties.cellWidth + this.carouselProperties.margin;
    }
    get visibleWidth() {
        return (this.carouselProperties.visibleWidth ||
            this.carouselProperties.cellsElement.parentElement.clientWidth);
    }
    updateProperties(carouselProperties) {
        this.carouselProperties = carouselProperties;
    }
    getStartX(event) {
        const { touches } = event;
        const carouselElementPosition = this.getCarouselElementPosition().left;
        let startX;
        if (touches) {
            startX = touches[0].clientX - carouselElementPosition;
        }
        else {
            startX = event.clientX - carouselElementPosition;
        }
        return startX;
    }
    getMoveX(event) {
        const { touches } = event;
        const carouselElementPositionX = this.getCarouselElementPosition().left;
        if (touches) {
            return touches[0].clientX - carouselElementPositionX;
        }
        return event.clientX - carouselElementPositionX;
    }
    getCarouselElementPosition() {
        return this.carouselProperties.hostElement.getBoundingClientRect();
    }
}
